README.md

# Flutter 加载 Html Flutter与JS的双向调用
Flutter是谷歌推出的最新的移动开发框架。


[【x1】微信公众号的每日提醒  随时随记 每日积累 随心而过](https://mp.weixin.qq.com/mp/homepage?__biz=MzA4MjIzOTE2Mw==&hid=1&sn=432233d701e1e1a93d897073955bd472&scene=18)  

[【x2】各种系列的视频教程 免费开源  关注 你不会迷路](https://www.toutiao.com/c/user/token/MS4wLjABAAAAYMrKikomuQJ4d-cPaeBqtAK2cQY697Pv9xIyyDhtwIM/)

[【x3】系列文章 百万 Demo 随时 复制粘贴 使用](https://biglead.blog.csdn.net/article/details/93532582)

***
大家如果有问题可以随时关注公众号联系作者，沟通解决需求。



  
>2023年4月30日 更新 1.3.2  Andorid 支持设置webView的缓存模式
```dart
enum WebViewCacheMode {
  LOAD_CACHE_ONLY, // 不发网络请求资源，只读取缓存。
  LOAD_DEFAULT, //根据cache-control或者Last-Modified决定是否从网络上取数据。默认采用该方案
  LOAD_NO_CACHE, //不使用缓存，只从网络获取数据。
  LOAD_CACHE_ELSE_NETWORK //只要本地有，无论是否过期，或者no-cache，都使用缓存中的数据。本地没有缓存时才从网络上获取。
}
```

* 2020-10-28 更新[1.1.4 版本 处理 Flutter与Html的混合页面加载](https://biglead.blog.csdn.net/article/details/109340398)

* 2020-10-27 更新 [1.1.3 版本](https://pub.flutter-io.cn/packages/flutter_fai_webview/install) 优化处理在 Android平台加载显示短暂的黑屏问题

* 2020-7-26 更新支持 Android iOS 平台向前或者是后退浏览历史功能
请查看本文章第四小节 

* 2020-8-12 修复在iphone8早JS调用Flutter失效问题[版本1.1.1](https://pub.flutter-io.cn/packages/flutter_fai_webview/install)
***
> 本篇文章讲述的内容可以用来加载 Html 页面，以实现 Android 中 WebView 或者 是 iOS 中的 UIWebView 中的功能。


**Flutter_Fai_Webview 插件可实现的功能：**

* 同时适配于 Android  Ios 两个平台
* 通过  url 来加载渲染一个Html 页面
* 加载 Html 文本数据 如 ```<html> .... </html>```等
* 加载 Html 标签数据 如  ```<p> ... </p>   ```
* 实现 WebView 加载完成后，自动测量 WebView 的高度，并回调 Flutter
* 实现 WebView 加载完成监听
* 实现 WebView 上下滑动、滑动到顶部兼听、滑动到底部兼听并回调 Flutter
* 实现 兼听 WebView 输出日志并将日志回调 Flutter 
* 实现 为 Html 页面中所有的图片添加点击事件 并回调 Flutter 
*  实现Html与Flutter的JS双向互调
* 实现打开相机相册的功能
* 实现回退历史浏览记录的功能
* 实现监听Html中图片的点击事件回调
* 刷新Html页面的加载

***

**Flutter中可用于来加载 Html 页面的插件 ,**

* flutter_WebView_plugin 
* webView_flutter
* flutter_inappbrowser 
* html 
* flutter_html 
* flutter_html_view 

这些多多少满足不了我项目中的需求，所以花了几天时间开发了 Flutter_Fai_Webview 插件，可实现   Android 中 WebView 或者 是 iOS 中的 UIWebView 中的功能，因为 Flutter_Fai_Webview 插件本质上是通过 PlatformView 功能将原生的 View 嵌套在 Flutter 中。

[插件源码在这里](https://github.com/zhaolongs/Flutter_Fai_Webview)

**开发插件要具备的知识：**

* Flutter 与 原生 Android iOS 双向通信 
 [Flutter通过MethodChannel实现Flutter 与Android iOS 的双向通信](https://blog.csdn.net/zl18603543572/article/details/96049359)
 [Flutter通过BasicMessageChannel实现Flutter 与Android iOS 的双向通信](https://blog.csdn.net/zl18603543572/article/details/96043692)
* Flutter 中内嵌 Android iOS 原生View的编程基础
	 [flutter调用android 原生TextView ](https://blog.csdn.net/zl18603543572/article/details/95983215)
	  [flutter调用ios 原生View ](https://blog.csdn.net/zl18603543572/article/details/96125516)
* 最重要的一点是 具备 Android  iOS 原生语言的开发能力



*** 

#### 开始使用

#### 1 基本使用说明
##### 1.1 Flutter 项目中 pubspec.xml 文件中 配置插件

pub方式依赖：[点击这里查看最新版本](https://pub.flutter-io.cn/packages/flutter_fai_webview)

```java
dependencies:
  flutter_fai_webview: ^1.0.0
```

git方式依赖：
```dart
  flutter_fai_webview:
    git:
      url: https://github.com/zhaolongs/Flutter_Fai_Webview.git
      ref: master
```

##### 1.2 在使用到 WebView 页面中 

引入头文件

```dart 
import 'package:flutter_fai_webview/flutter_fai_webview.dart';
```

##### 1.3 通过url加载网页
如这里使用到的地址：

```java
  String htmlUrl = "https://blog.csdn.net/zl18603543572";
```
然后使用FaiWebViewWidget来加载显示这个h5链接，代码如下：


```java 

  ///通过url加载页面
  FaiWebViewWidget buildFaiWebViewWidget() {
    return FaiWebViewWidget(
      //webview 加载网页链接
      url: htmlUrl,
      //webview 加载信息回调
      callback: callBack,
      //输出日志
      isLog: true,
    );
  }
```
其中callBack是一个回调，如webview的加载完成、向上滑动、向下滑动等等，代码如下：

```java
  ///加载 Html 的回调
  ///[code]消息类型标识
  ///[msg] 消息内容
  ///[content] 回传的参数
  void callBack(int ?code, String ?msg, content) {
    //加载页面完成后 对页面重新测量的回调
    //这里没有使用到
    //当FaiWebViewWidget 被嵌套在可滑动的 widget 中，必须设置 FaiWebViewWidget 的高度
    //设置 FaiWebViewWidget 的高度 可通过在 FaiWebViewWidget 嵌套一层 Container 或者 SizeBox
    if (code == 201) {
      //页面加载完成后 测量的 WebView 高度
      double webViewHeight = content;
      print("webViewHeight " + webViewHeight.toString());
    } else {
      //其他回调
    }
    setState(() {
      message = "回调：code[" + code.toString() + "]; msg[" + msg.toString() + "]";
    });
  }
```
运行效果如下图所示：

![在这里插入图片描述](https://img-blog.csdnimg.cn/20200714211841289.gif)
[完整代码在这里](https://github.com/zhaolongs/Flutter_Fai_Webview/blob/master/example/lib/exampl_default_url_max.dart)
##### 1.4 关于回调中的code取值说明如下
  
| code值 |  简述|
|--|--|
| 201 |测量webview 成功
| 202 |JS调用
| 203 |图片点击回调
|301 |滑动到顶部
|302 |向下滑动
| 303|向上滑动
| 304 |滑动到底部
| 401 |webview 开始加载
|402 |webview 加载完成
|403 |webview html中日志输出
| 404 |webview 加载出错
| 501 |webview 弹框回调
|1000 |操作失败

加载Html的过程中会实时回调Flutter，详细说明如下：
```
  /**
   * code 原生 Android iOS 回调 Flutter 的消息类型标识
   * message 消息类型日志
   * content 回调的基本数据
   */
  Function(int code, String message, dynamic content) callback;
```

详细说明

```java

  //当前点击的图片 URL
  String imageUrl = null;
  //是否显示浮动按钮
  bool isShowFloat = false;
  /**
   * code 当前点击图片的 位置
   * url 当前点击图片对应的 链接
   * images 当前 Html 页面中所有的图片集合
   */
  void imageCallBack(int code, String url, List<String> images) {
    imageUrl = url;
    setState(() {});
  }

  void callBack(int ?code, String ?msg, content) {
    String call = "回调 code:" +
        code.toString() +
        " msg:" +
        msg.toString() +
        " content:" +
        content.toString();
    if (code == 201) {
      //加载页面完成后 对页面重新测量的回调
      //这里没有使用到
      //当FaiWebViewWidget 被嵌套在可滑动的 widget 中，必须设置 FaiWebViewWidget 的高度
      //设置 FaiWebViewWidget 的高度 可通过在 FaiWebViewWidget 嵌套一层 Container 或者 SizeBox
      webViewHeight = content;
    } else if (code == 202) {
      // Html 页面中 Js 的回调
      // Html 页面中的开发需要使用 Js 调用  【 Android 中 使用 controll.otherJsMethodCall( json )】 【iOS中 直接调用 otherJsMethodCall( json ) 】
      // 在 Flutter 中解析 json 然后加载不同的功能
      String jsJson = content;
    } else if (code == 203) {
      // 为 Html 页面中的图片添加 点击事件后，点击图片会回调此方法
      // content 为当前点击图片的 地址
      // 实现更多功能 比如 一个 Html 页面中 有5张图片，点击放大查看并可右右滑动
      // 这个功能可以在 imageCallBack 回调中处理

    } else if (code == 301) {
      //当 WebView 滑动到顶部的回调
    } else if (code == 302) {
      //当 WebView 开始向下滑动时的回调
      //隐藏按钮
      isShowFloat = true;
    } else if (code == 303) {
      //当 WebView 开始向上滑动时的回调
      //显示按钮
      isShowFloat = false;
    } else if (code == 304) {
      //当 WebView 滑动到底部的回调
    } else if (code == 401) {
      //当 WebView 开始加载的回调
    } else if (code == 402) {
      //当 WebView 加载完成的回调
    } else if (code == 403) {
      // WebView 中 Html中日志输出回调
    } else if (code == 401) {
      // WebView 加载 Html 页面出错的回调
    } else if (code == 501) {
      // 当 Html 页面中有 Alert 弹框弹出时 回调消息

    } else if (code == 1000) {
      // 操作失败 例如 空指针异常 等等
    } else {
      //其他回调
    }
    setState(() {
      message = call;
    });
  }
```

***
![公众号 我的大前端生涯](https://img-blog.csdnimg.cn/20200620175409480.gif)
***

 ####  2 Flutter 加载页面 
 
##### 2.1  通过 url 加载 Html 页面
在上述1.3 通过url加载网页已进行过描述

##### 2.2 通过 Html Data 加载 String类型的Html 页面
加载String类型的Html页面，一般先是将String类型的Html代码加载到内容中，如通过网络请求接口获取的，或者是在页面中定义好的代码如下:

```java
 String htmlBlockData =
      "<!DOCTYPE html><html> <head><meta http-equiv=\"Content-Type\" content=\"text/html; charset=UTF-8\">  <meta name=\"viewport\" content=\"width=device-width,initial-scale=1,maximum-scale=1\"> </head> <body><p>加载中</p></body></html>";

```

或者是放在assets目录的静态Html，那么就需要先读取静态资源目录下的Html文件 ，代码如下：

```java
  ///读取静态Html
  Future<String> loadingLocalAsset() async {
    ///加载
    String htmlData = await rootBundle.loadString('assets/html/test.html');
    print("加载数据完成 $htmlData");
    return htmlData;
  }
```
当然你需要配制好目录依赖，如我这里的example中的html文件配置代码如下：

```java
  assets:
    - assets/html/
```
对应的文件目录如下所示：
![](https://img-blog.csdnimg.cn/20200714213210344.png?x-oss-process=image/watermark,type_ZmFuZ3poZW5naGVpdGk,shadow_10,text_aHR0cHM6Ly9ibG9nLmNzZG4ubmV0L3psMTg2MDM1NDM1NzI=,size_16,color_FFFFFF,t_70)
然后就是使用FaiWebViewWidget来渲染Html页面了，如果是直接显示已加载好的String类型的Html，那么直接配置FaiWebViewWidget中的htmlBlockData就可以，如果是使用异步加载assets目录下的静态Html，那么可以结合FutureBuilder组件来实现加载，代码如下：

```java
  ///异步加载静态资源目录下的Html
  FutureBuilder<String> buildFutureBuilder() {
    return FutureBuilder<String>(
      ///异步加载数据
      future: loadingLocalAsset(),
      ///构建
      builder: (BuildContext context, var snap) {
        ///加载完成的html数据
        String htmlData = snap.data;
        //使用插件 FaiWebViewWidget
        if (htmlData == null) {
          return CircularProgressIndicator();
        }
        ///通过配置 htmlBlockData 来渲染
        return FaiWebViewWidget(
          //webview 加载本地html数据
          htmlBlockData: htmlData,
          //webview 加载信息回调
          callback: callBack,
          //输出日志
          isLog: true,
        );
      },
    );
  }
```
[对应的完整代码在这里](https://github.com/zhaolongs/Flutter_Fai_Webview/blob/master/example/lib/exampl_default_html_data_max_refresh2.dart)

##### 2.3 加载混合页面
也就是说 一个页面中，一部分是 Flutter Widget 一部分是 webview 加载。

![在这里插入图片描述](https://img-blog.csdnimg.cn/20190720154728780.gif)

```java
import 'package:flutter/material.dart';
import 'dart:async';
import 'package:flutter_fai_webview/flutter_fai_webview.dart';

/**
 *   混合页面加载
 *
 */
class DefaultHexRefreshPage extends StatefulWidget {
  @override
  MaxUrlHexRefreshState createState() => MaxUrlHexRefreshState();
}

class MaxUrlHexRefreshState extends State<DefaultHexRefreshPage> {
 

  //原生 发送给 Flutter 的消息
  String message = "--";
  String htmlUrl = "https://blog.csdn.net/zl18603543572";
  double webViewHeight = 1;

  @override
  void initState() {
    super.initState();

  }

  @override

CHANGELOG.md
## 1.3.3 
 Android 支持清除历史记录 、清除缓存的功能

```dart

FaiWebViewController faiWebViewController = new FaiWebViewController();

//点击按钮调用
faiWebViewController.clearHistory();
faiWebViewController.clearCache();

//需要注意的是 WebView 需要绑定 

FaiWebViewWidget(
    controller: faiWebViewController,
    ... 
)


```
## 1.3.2 
 Andorid 支持设置webView的缓存模式
```dart
enum WebViewCacheMode {
  LOAD_CACHE_ONLY, // 不发网络请求资源，只读取缓存。
  LOAD_DEFAULT, //根据cache-control或者Last-Modified决定是否从网络上取数据。默认采用该方案
  LOAD_NO_CACHE, //不使用缓存，只从网络获取数据。
  LOAD_CACHE_ELSE_NETWORK //只要本地有，无论是否过期，或者no-cache，都使用缓存中的数据。本地没有缓存时才从网络上获取。
}
```

## 1.3.1

设置WebView的缓存策略

## 1.3.0

* notsafenull
* 新增 WebView 添加 脚视图的功能
* 新增自定义webView高度的功能

```
Container buildFaiWebViewWidget() {
  return Container(
    child: FaiWebViewWidget(
      ///下拉刷新
      onRefresh: _onRefresh,
      //头
      headerWidget: buildHeaderWidget(),
      //脚
      footerWidget: buildHeaderWidget(),
      //WebView的高度
      //webViewHeight: 400,
      //webview 加载网页链接
      url: htmlUrl,
      //webview 加载信息回调
      callback: callBack,
      //输出日志
      isLog: false,
    ),
  );
}

```
## 1.1.5

* 添加 加载中显示文案
* 兼容部分机型有底部虚拟菜单栏无法监听回调问题
* 优化 ScrollView 与 WebView 的兼容显示

## 1.1.4

* 添加 Scrollview 与webview 的混合加载使用

## 1.1.2

* 添加在加载 Html 字符串时 默认 编码为  utf-8 

## 1.1.1

* 修复在 iphone 8 下  JS 回调 Flutter 无效问题

## 1.1.0 

* 新增的对浏览网页历史的前进与后退  Android iOS 都支持

## 1.0.0 

* 完善了Flutter与HTML中的JS的双向互调

## 0.0.2

* 添加加载资源目录assets下静态html文件

## 0.0.1

* TODO: Describe initial release.




lib/src/native_webview.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_fai_webview/src/fai_webview_controller.dart';

/// 向 Flutter 中发送消息
/// code
/// 201 测量webview 成功
/// 202 JS调用
/// 203 图片点击回调
/// 301 滑动到顶部
/// 302 向下滑动
/// 303	向上滑动
/// 304 滑动到底部
/// 401 webview 开始加载
/// 402 webview 加载完成
/// 403 webview html中日志输出
/// 404 webview 加载出错
/// 501 webview 弹框回调
/// <p>
/// 1000 操作失败
///
/// @param map
/// code :203 图片点击回调
/// // url 当前点击图片的链接 index 当前点击Html页面中所有图片中的角标 urls 所有图片的集合
/// content: {"url":"http://pic.studyyoun.com/1543767087584","index":0,"urls":"http://pic.studyyoun.com/1543767087584,http://pic.studyyoun.com/1543767100547"}

enum WebViewCacheMode {
  LOAD_CACHE_ONLY, // 不发网络请求资源，只读取缓存。
  LOAD_DEFAULT, //根据cache-control或者Last-Modified决定是否从网络上取数据。默认采用该方案
  LOAD_NO_CACHE, //不使用缓存，只从网络获取数据。
  LOAD_CACHE_ELSE_NETWORK //只要本地有，无论是否过期，或者no-cache，都使用缓存中的数据。本地没有缓存时才从网络上获取。
}

class FaiWebViewWidget extends StatefulWidget {
  //加载的网页 URL
  final String? url;

  //加载 完整html 文件数据 如 <html><head> .... .. </head></html>
  final String? htmlData;

  //加载 html 代码块 如<p> .... </p>
  final String? htmlBlockData;

  //日志输出
  final bool isLog;

  //HTML中的图片添加点击事件
  final bool htmlImageIsClick;

  final WebViewCacheMode webViewCacheMode;

  ///下拉刷新回调
  final Function? onRefresh;

  /// [code]  原生 Android iOS 回调 Flutter 的消息类型标识
  /// [message]  消息类型日志
  /// [content]  回调的基本数据
  final Function(int? code, String? message, dynamic content)? callback;

  /// 图片点击回调
  ///[index] HTML 中图片索引
  /// [url] 当前点击的图片的地址
  /// [images] HTML中所有的图片的集合
  final Function(int index, String url, List<String> images)? imageCallBack;

  ///操作webView的控制器
  final FaiWebViewController? controller;

  final ScrollController? scrollController;

  final Widget? headerWidget;
  final Widget? footerWidget;
  final Widget? appBar;
  final double? webViewHeight;
  final double? minHeight;

  FaiWebViewWidget(
      {
      //webview 加载网页链接
      this.url,
      //webview 加载 完整的 html 文件数据  如 <html> .... </html>
      // 不完整的 html 文件数据 如 <p></p> 配置到此项，用此属性来加载，只会渲染 <p> ... </p> 中已有的样式 不会适配移动端显示
      this.htmlData,
      //webview 加载完整的 html 文件数据 或者是 不完整的 html 文件数据 如 <p></p>
      //不完整的 html 文件数据 如 <p></p> 配置到此项，会自动将不完整的 html 文件数据 添加 <html><head> .. </head> <body> 原来的内容 </body></html>,并适配移动端
      this.htmlBlockData,
      //输出 Log 日志功能
      this.isLog = false,
      this.webViewCacheMode = WebViewCacheMode.LOAD_DEFAULT,
      // 为 Html 页面中所有的图片添加 点击事件 并通过回调 通知 Flutter 页面
      // 只有使用 htmlBlockData 属性加载的页面才会有此效果
      this.htmlImageIsClick = false,
      // Html 页面中图片点击回调
      this.imageCallBack,
      // Html 页面中所有的消息回调
      this.callback,
      this.controller,
      this.scrollController,
      //混合加载时 WebView 顶部的 Widget
      this.headerWidget,
      this.onRefresh,
      this.appBar,
      //是否显示默认的加载中
      this.showLoading = true,
      //自定义加载中 Widget
      this.loadginWidget,
      //WebView 下面的 Widget
      this.footerWidget,
      //webView 的高度 如果指定了 会使用这里的值 如果未指定 会自动测量
      this.webViewHeight,
      this.minHeight,
      Key? key})
      : super(key: key);

  final bool showLoading;

  final Widget? loadginWidget;

  @override
  State<StatefulWidget> createState() {
    return AndroidWebViewState();
  }
}

class AndroidWebViewState extends State<FaiWebViewWidget> {
  double? _webViewHeight;
  double _defaultWebViewHeight = 40;

  ///滑动布局控制器
  ScrollController _scrollController = new ScrollController();

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    _webViewHeight = widget.webViewHeight;
  }

  @override
  void didUpdateWidget(covariant FaiWebViewWidget oldWidget) {
    // TODO: implement didUpdateWidget
    super.didUpdateWidget(oldWidget);
    _webViewHeight = widget.webViewHeight;
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: widget.appBar as PreferredSizeWidget?,
      body: buildRefreshHexWidget(),
    );
  }

  //double _dy = 0.0;

  findCurrentDy() {
    RenderBox? findRenderObject = context.findRenderObject() as RenderBox?;
    if (findRenderObject != null) {
      //Offset localOffset = findRenderObject.localToGlobal(Offset.zero);
      // _dy = localOffset.dy;
    }
  }

  /// 需要注意的是
  /// RefreshIndicator 会覆盖 WebView 的滑动事件
  /// 所有关于 监听 WebView 的滑动监听将会失效
  Widget buildRefreshHexWidget() {
    Widget itemWidget = SingleChildScrollView(
      controller: _scrollController,
      physics: _scrollPhysics,
      child: buildColumn(),
    );

    if (widget.onRefresh != null) {
      itemWidget = RefreshIndicator(
        //下拉刷新触发方法
        onRefresh: widget.onRefresh as Future<void> Function(),
        //设置webViewWidget
        child: itemWidget,
      );
    }

    return SizedBox.expand(
      child: NotificationListener(
        onNotification: (ScrollNotification notification) {
          //滑动信息处理
          //根据不同的滑动信息来处理页面的特效
          //如Widget 移动 、放大、缩小、旋转等等
          notificationFunction(notification);
          //可滚动组件在滚动过程中会发出ScrollNotification之外，
          //还有一些其它的通知，
          //如SizeChangedLayoutNotification、
          //   KeepAliveNotification 、
          //   LayoutChangedNotification等
          //返回值类型为布尔值，当返回值为true时，阻止冒泡，
          //其父级Widget将再也收不到该通知；当返回值为false 时继续向上冒泡通知。
          return true;
        },
        child: itemWidget,
      ),
    );
  }

  notificationFunction(ScrollNotification notification) {
    //滑动信息的数据封装体
    ScrollMetrics metrics = notification.metrics;
    //当前位置
    double pixels = metrics.pixels;
    print("当前位置 $pixels");

    //滚动类型
    Type runtimeType = notification.runtimeType;

    switch (runtimeType) {
      case ScrollStartNotification:
        print("开始滚动");
        break;
      case ScrollUpdateNotification:
        print("正在滚动");
        break;
      case ScrollEndNotification:
        print("滚动停止");
        double offset = _scrollController.offset;
        if (widget.footerWidget == null &&
            offset == _scrollController.position.maxScrollExtent) {
          print("滚动停止  _scrollPhysics");
          _scrollPhysics = NeverScrollableScrollPhysics();
          setState(() {});
        }
        break;
      case OverscrollNotification:
        break;
    }
  }

  Widget buildColumn() {
    if (widget.headerWidget == null && widget.footerWidget == null) {
      return buildContainer();
    } else {
      return Column(
        children: [
          buildHeader(),
          buildContainer(),
          buildFoot(),
        ],
      );
    }
  }

  buildHeader() {
    if (widget.headerWidget == null) {
      return Container();
    }
    return widget.headerWidget!;
  }

  buildFoot() {
    if (widget.footerWidget == null) {
      return Container();
    }
    return widget.footerWidget!;
  }

  Container buildContainer() {
    double? height =
        _webViewHeight == null ? _defaultWebViewHeight : _webViewHeight;
    return Container(
      height: height,
      child: Stack(
        children: [
          buildFaiWebViewItemWidget(),
          buildLoadingWidget(),
        ],
      ),
    );
  }

  Widget buildLoadingWidget() {
    if (widget.showLoading && !isHideLoading) {
      if (widget.loadginWidget == null) {
        return Center(
          child: Text("加载中..."),
        );
      } else {
        return widget.loadginWidget!;
      }
    } else {
      return Container();
    }
  }

  FaiWebViewItemWidget buildFaiWebViewItemWidget() {
    return FaiWebViewItemWidget(
      //webview 加载网页链接
      url: widget.url,
      webViewCacheMode: widget.webViewCacheMode,
      htmlData: widget.htmlData,
      htmlBlockData: widget.htmlBlockData,
      htmlImageIsClick: widget.htmlImageIsClick,
      //webview 加载信息回调
      callback: callBack,
      //输出日志
      isLog: widget.isLog,
      controller: widget.controller,
    );
  }

  ScrollPhysics _scrollPhysics = ClampingScrollPhysics();

  ///是否需要滑动兼容处理
  bool isScrollHex = false;

  bool isHideLoading = false;

  callBack(int? code, String? msg, content) {
    //加载页面完成后 对页面重新测量的回调
    if (code == 201) {
      double widgetPerentHeight = MediaQuery.of(context).size.height * 3;
      findCurrentDy();
      //double flagHeight = widgetPerentHeight;
      if (_webViewHeight == null) {
        if (content <= widgetPerentHeight) {
          _webViewHeight = content;
          if (widget.minHeight != null && _webViewHeight! < widget.minHeight!) {
            _webViewHeight = widget.minHeight;
          }
          isScrollHex = false;
        } else {
          _webViewHeight = widgetPerentHeight;
          isScrollHex = true;
        }
      }
      isHideLoading = true;
      //更新高度
      setState(() {});
      print("webViewHeight " + content.toString());
    } else if (code == 301 && widget.headerWidget != null) {
      print("_scrollPhysics 更新");
      //其他回调
      _scrollPhysics = ClampingScrollPhysics();
      setState(() {});
    }
    if (widget.callback != null) {
      widget.callback!(code, msg, content);
    }
  }
}

class FaiWebViewItemWidget extends StatefulWidget {
  //加载的网页 URL
  final String? url;

  //加载 完整html 文件数据 如 <html><head> .... .. </head></html>
  final String? htmlData;

  //加载 html 代码块 如<p> .... </p>
  final String? htmlBlockData;

  //日志输出
  final bool isLog;

  //HTML中的图片添加点击事件
  final bool htmlImageIsClick;

  /// [code]  原生 Android iOS 回调 Flutter 的消息类型标识
  /// [message]  消息类型日志
  /// [content]  回调的基本数据
  final Function(int? code, String? message, dynamic content)? callback;

  /// 图片点击回调
  ///[index] HTML 中图片索引
  /// [url] 当前点击的图片的地址
  /// [images] HTML中所有的图片的集合
  final Function(int? index, String? url, List<String>? images)? imageCallBack;

  ///操作webView的控制器
  final FaiWebViewController? controller;

  final WebViewCacheMode webViewCacheMode;

  FaiWebViewItemWidget({
    //webview 加载网页链接
    this.url,
    //webview 加载 完整的 html 文件数据  如 <html> .... </html>
    // 不完整的 html 文件数据 如 <p></p> 配置到此项，用此属性来加载，只会渲染 <p> ... </p> 中已有的样式 不会适配移动端显示
    this.htmlData,
    //webview 加载完整的 html 文件数据 或者是 不完整的 html 文件数据 如 <p></p>
    //不完整的 html 文件数据 如 <p></p> 配置到此项，会自动将不完整的 html 文件数据 添加 <html><head> .. </head> <body> 原来的内容 </body></html>,并适配移动端
    this.htmlBlockData,
    //输出 Log 日志功能
    this.isLog = false,
    // 为 Html 页面中所有的图片添加 点击事件 并通过回调 通知 Flutter 页面
    // 只有使用 htmlBlockData 属性加载的页面才会有此效果
    this.htmlImageIsClick = false,
    // Html 页面中图片点击回调
    this.imageCallBack,
    // Html 页面中所有的消息回调
    this.callback,
    this.controller,
    this.webViewCacheMode = WebViewCacheMode.LOAD_DEFAULT,
  });

  @override
  State<StatefulWidget> createState() {
    return FaiWebViewItemWidgetState();
  }
}

class FaiWebViewItemWidgetState extends State<FaiWebViewItemWidget> {
  //自定义网页中的所有的图片的点击事件处理

  //加载 完整html 文件数据 如 <html><head> .... .. </head></html>

  //加载 html 代码块 如<p> .... </p>

  //日志输出

  int viewId = -1;
  late MethodChannel _channel;

  String? mUrl;

  String? mHtmlData;
  String? mHtmlBlockData;

  @override
  void initState() {
    super.initState();
    mUrl = widget.url;
    mHtmlData = widget.htmlData;
    mHtmlBlockData = widget.htmlBlockData;
    if (widget.controller != null) {
      widget.controller!.setListener(webViewListener);
      widget.controller!.setBackListener(webViewBackListener);
    }
  }

  @override
  Widget build(BuildContext context) {
    if (Platform.isIOS) {
      //ios相关代码
      return buildIosWebView();
    } else if (Platform.isAndroid) {
      //android相关代码
      return buildAndroidWebView();
    } else {
      return Container(
        child: Text("暂不支持当前平台"),
      );
    }
  }

  Future<bool?> webViewBackListener(int type, Map<String, dynamic>? event) {
    if (type == 1) {
      return canGoBack();
    } else if (type == 2) {
      return canGoForward();
    } else {
      return Future.value(false);
    }
  }

  void webViewListener(type, event) {
    ///刷新页面方法
    ///目前1.0.0版本还不可动态修改内容
    if (type == 1) {
      String? htmlData = event["htmlData"];
      String? htmlBlockData = event["htmlBlockData"];
      String? htmlUrl = event["htmlUrl"];

      if (htmlData != null) {
        mHtmlData = htmlData;
      }
      if (htmlBlockData != null) {
        mHtmlBlockData = htmlBlockData;
      }
      if (htmlUrl != null) {
        mUrl = htmlUrl;
      }
      refresh();
    } else if (type == 2) {
      String? jsMethodName = event["jsMethodName"];

      ///Flutter调用 Html中的Js方法
      loadJsMethod("$jsMethodName('${json.encode(event)}')");
    } else if (type == 3) {
      ///返回历史
      goBack();
    } else if (type == 4) {
      ///返回历史
      goForward();
    }else if (type == 5) {
      ///清除历史
      goClearHistory();
    }else if (type == 6) {
      ///清除历史
      goClearCache();
    }
  }

  @override
  void dispose() {
    super.dispose();
    // NativeEventMessage.getDefault().unregister();
  }

  /// 设置消息监听
  /// code
  /// 201 测量webview 成功 加载完成
  /// 202 JS调用
  /// 301 滑动到顶部
  /// 302 向下滑动
  /// 303	向上滑动
  /// 304 滑动到底部
  /// 401 webview 开始加载
  /// 402 webview 加载完成
  /// 403 webview html中日志输出
  /// 404 webview 加载出错
  ///
  /// @param map
  void nativeMessageListener() async {
    _channel.setMethodCallHandler((resultCall) {
      //处理原生 Android iOS 发送过来的消息
      MethodCall call = resultCall;
      //String method = call.method;
      Map arguments = call.arguments;

      int? code = arguments["code"];
      String? message = arguments["message"];
      dynamic content = arguments["content"];
      print("native_webview:code-> " +
          code.toString() +
          " ; message:" +
          message.toString() +
          "; content " +
          content.toString());

      if (code == 203) {
        int? index = arguments["index"];
        String? url = arguments["url"];
        List<String>? urls = arguments["urls"];
        if (widget.imageCallBack != null) {
          widget.imageCallBack!(index, url, urls);
        }
      } else if (code == 201) {
        _streamController.add(1.0);
      }

      if (widget.callback != null) {
        print("native_webview callback");
        widget.callback!(code, message, content);
      } else {
        print("native_webview callback is null ");
      }
      return Future.value(true);
    });
  }

  void loadUrl() async {
    _channel.invokeMethod('load', {
      "url": mUrl,
      "htmlData": mHtmlData,
      "htmlBlockData": mHtmlBlockData,
    });
  }

  void reLoad() async {
    _channel.invokeMethod('reload');
  }

  void goBack() {
    _channel.invokeMethod('goBack');
  }

  void goForward() {
    _channel.invokeMethod('goForward');
  }
  void goClearHistory(){
    _channel.invokeMethod('goClearHistory');
  }
  void goClearCache(){
    _channel.invokeMethod('goClearCache');
  }


  Future<bool?> canGoBack() async {
    if (Platform.isAndroid) {
      return await _channel.invokeMethod('canGoBack');
    } else {
      return await _channel.invokeMethod('canGoBack') == "false" ? false : true;
    }
  }

  Future<bool?> canGoForward() async {
    if (Platform.isAndroid) {
      return await _channel.invokeMethod('canGoForward');
    } else {
      return await _channel.invokeMethod('canGoForward') == "false"
          ? false
          : true;
    }
  }

  StreamController<double> _streamController = new StreamController();

  /// 监听Stream，每次值改变的时候，更新Text中的内容
  StreamBuilder<double> buildAndroidWebView() {
    return StreamBuilder<double>(
      //绑定stream
      stream: _streamController.stream,
      //默认的数据
      initialData: 0.0,
      //构建绑定数据的UI
      builder: (BuildContext context, AsyncSnapshot<double> snapshot) {
        return buildStreamBuilder(snapshot.data!);
      },
    );
  }

  Widget buildStreamBuilder(double data) {
    return Opacity(
      opacity: data,
      child: AndroidView(
        //调用标识
        viewType: "com.flutter_to_native_webview",
        //参数初始化
        creationParams: {
          "isScrollListen": true,
          "htmlImageIsClick": widget.htmlImageIsClick,
          "webViewCacheMode":widget.webViewCacheMode.toString(),
        },
        //参数的编码方式
        creationParamsCodec: const StandardMessageCodec(),
        //webview 创建后的回调
        onPlatformViewCreated: (id) {
          platformViewCreatedFunction(id);
        },
      ),
    );
  }

  Widget buildIosWebView() {
    return UiKitView(
      //调用标识
      viewType: "com.flutter_to_native_webview",
      //参数初始化
      creationParams: {
        //调用view参数标识
        "isScrollListen": true,
        "htmlImageIsClick": widget.htmlImageIsClick
      },
      //参数的编码方式
      creationParamsCodec: const StandardMessageCodec(),
      //webview 创建后的回调
      onPlatformViewCreated: (id) {
        platformViewCreatedFunction(id);
      },
    );
  }

  void platformViewCreatedFunction(int id) {
    viewId = id;
    print("onPlatformViewCreated " + id.toString());
    //创建通道
    _channel = new MethodChannel('com.flutter_to_native_webview_$viewId');
    //设置监听
    nativeMessageListener();
    //加载页面
    loadUrl();
  }

  void refresh() {
    reLoad();
  }

  void loadJsMethod(String string) async {
    _channel.invokeMethod('jsload', {
      "string": string,
    });
  }
}
